gnovm/pkg/gnolang/type_check.go

```diff
@@ -0,0 +1,892 @@
+		}
+		return
+	case *ArrayType:
+		if at, ok := xt.(*ArrayType); ok {
+			checkAssignableTo(at.Elt, cdt.Elt, false)
+			return
+		}
+	case *SliceType:
+		if st, ok := xt.(*SliceType); ok {
+			checkAssignableTo(st.Elt, cdt.Elt, false)
+			return
+		}
+	case *MapType:
+		if mt, ok := xt.(*MapType); ok {
+			checkAssignableTo(mt.Key, cdt.Key, false)
+			checkAssignableTo(mt.Value, cdt.Value, false)
+			return
+		}
+	case *FuncType:
+		if xt.TypeID() == cdt.TypeID() {
+			return // ok
+		}
+	case *InterfaceType:
+		panic("should not happen")
+	case *DeclaredType:
+		// do nothing, untyped to declared type
+		return
+		// panic("should not happen")
+	case *StructType, *PackageType, *ChanType:
+		if xt.TypeID() == cdt.TypeID() {
+			return // ok
+		}
+	case *TypeType:
+		if xt.TypeID() == cdt.TypeID() {
+			return // ok
+		}
+	case *NativeType:
+		if !autoNative {
+			if debug {
+				debug.Printf("native type, xt.TypeID: %v, cdt.TypeID: %v \n", xt.TypeID(), cdt.TypeID())
+			}
+			if xt.TypeID() == cdt.TypeID() {
+				return // ok
+			}
+		} else {
+			// autoNative, so check whether matches.
+			// xt: any type but a *DeclaredType; could be native.
+			// cdt: actual concrete native target type.
+			// ie, if cdt can match against xt.
+			if gno2GoTypeMatches(xt, cdt.Type) {
+				return // ok
+			}
+		}
+	default:
+		panic(fmt.Sprintf(
+			"unexpected type %s",
+			dt.String()))
+	}
+	panic(fmt.Sprintf(
+		"cannot use %s as %s",
+		xt.String(),
+		dt.String()))
+}
+
+// ===========================================================
+func (bx *BinaryExpr) checkShiftExpr(dt Type) {
+	var destKind interface{}
+	if dt != nil {
+		destKind = dt.Kind()
+	}
+	if checker, ok := binaryChecker[bx.Op]; ok {
+		if !checker(dt) {
+			panic(fmt.Sprintf("operator %s not defined on: %v", wordTokenStrings[bx.Op], destKind))
+		}
+	} else {
+		panic("should not happen")
+	}
+}
+
+// AssertCompatible works as a pre-check prior to checkOrConvertType()
+// It checks against expressions to ensure the compatibility between operands and operators.
+// e.g. "a" << 1, the left hand operand is not compatible with <<, it will fail the check.
+// Overall,it efficiently filters out incompatible expressions, stopping before the next
+// checkOrConvertType() operation to optimize performance.
+func (bx *BinaryExpr) AssertCompatible(lt, rt Type) {
+	// we can't check compatible with native types at current stage,
+	// so leave it to later operations(trans_leave on binaryExpr)
+	// to be converted into gno(only for primitive types), and do
+	// this check again. (AssertCompatible would be invoked again)
+	// non-primitive types is a special case that is not handled.
+	if lnt, ok := lt.(*NativeType); ok {
+		_, ok := go2GnoBaseType(lnt.Type).(PrimitiveType)
+		if ok {
+			return
+		}
+	}
+	if rnt, ok := rt.(*NativeType); ok {
+		_, ok := go2GnoBaseType(rnt.Type).(PrimitiveType)
+		if ok {
+			return
+		}
+	}
+
+	escapedOpStr := strings.Replace(wordTokenStrings[bx.Op], "%", "%%", 1)
```

Comment:
You don't need to do this, because the value is not interpolated into the format string, same goes for all occurrences

ie. you'd have to do this if you did something like `fmt.Sprintf("hello" + escapedOpStr)`; but you are doing `fmt.Sprintf("hello%v", escapedOpStr)`, so you don't actually need to escape it

Reply:
yeah this is a misjudgment, there's actually a bug in the render side at tm2/pkg/errors, I have a fix on that so there's no need to escape here.